### PR TITLE
automatic re-index of parent work when asset changes in relevant ways

### PR DESCRIPTION
**Pull request by jrochkind**
Ref #1428

While not a lot of code, this is some pretty "clever" code, dealing with some obscure parts of Kithe, attr_json, and ActiveRecord. I don't love it, but this is kind of the path we're on with our "automatic solr indexing". 

Tried to comment extensively. 

The ugliest part is that extra `if self.destroyed` line in the model. That one isn't actually covered by tests, I just found it trying out in the actual app, and then had to debug it, a real head-scratcher. 

Going to leave this in draft mode until next week, see if I can think of anything else.

### CODE REVIEW

**Comment by jrochkind** [comment]
It doesn't yet have a test for the thing I found last minute to fix in an even *more* confusing way, that deleting an asset was *not* causing re-indexing. 

i want to figure out a way to add an otherwise-failing test for that, otherwise it's done. With some sadness because it is, I realize, kind of convoluted (although concise) code.

**Comment by jrochkind** [comment]
Another thing that makes this all confusing is the code around whether when running in our tests we trigger automatic solr indexes. 

We ordinarily don't want to in tests. But here we want to TEST that, so sometimes have to turn it on.... but still don't want it turned on when we're doing our DB setup. Phew!

File: app/models/asset.rb [hunk 1]
```diff
@@ -58,6 +66,54 @@ def dzi_file
...
+  #   * update PARENT in solr on change, we don't index Assets, but do index
+  #     some aspects of assets in their parent Works.
+  #   * and only when allow-listed attributes we know we index on parent have changed
+  def update_index(mapper: kithe_indexable_mapper, writer:nil, **)
+    if should_reindex_parent_after_save?
+      # WEIRD workaround, in some cases the parent still this record in memory
+      # even though it's been, and will use that in-memory list of members for
+      # indexing in Solr! If that's going on, we need to make sure to reset
+      # the association.
+      if self.destroyed? && parent.members.loaded? && parent.members.include?(self)
```
Comment (eddierubeiz) [inline app/models/asset.rb:79]: Meaning -- in some cases the parent work still contains this asset in its `members` association even though the child asset has already been deleted, and could attempt to index information from the deleted asset.

File: app/models/asset.rb [hunk 2]
```diff
@@ -58,6 +66,54 @@ def dzi_file
...
+    elsif self.saved_change_to_published?
+      # if published status changed, we have to reindex if and only if we HAVE any indexed attributes,
+      # to include them in index.
+      indexed_attributes.any? { |attr| self.send(attr).present? }
+    else
+      # an ordinary save (including create), did any attributes of interest CHANGE? (Including removal)
+      # then we need to reindex parent to get them updated in index.
+      indexed_attributes.any? { |attr| self.attr_json_changes.saved_change_to_attribute(attr).present? }
+    end
+  end
```
Comment (eddierubeiz) [inline app/models/asset.rb:115]: OK, this all makes sense.

**Comment by eddierubeiz** [comment]
OK. The tests are invaluable here as documentation, but so are the comments. Merging!

**Pull request merged**